go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -0,0 +1,293 @@
+	defer cancel()
+
+	settings, err := binlogplayer.ReadVRSettings(vc.vr.dbClient, vc.vr.id)
+	if err != nil {
+		return err
+	}
+	if settings.StartPos.IsZero() {
+		return nil
+	}
+
+	// Start vreplication.
+	errch := make(chan error, 1)
+	go func() {
+		errch <- newVPlayer(vc.vr, settings, copyState, mysql.Position{}).play(ctx)
+	}()
+
+	// Wait for catchup.
+	tmr := time.NewTimer(1 * time.Second)
+	seconds := int64(replicaLagTolerance / time.Second)
+	defer tmr.Stop()
+	for {
+		sbm := vc.vr.stats.SecondsBehindMaster.Get()
+		if sbm < seconds {
+			cancel()
+			<-errch
+			return nil
+		}
+		select {
+		case err := <-errch:
+			if err != nil {
+				return err
+			}
+			return io.EOF
+		case <-ctx.Done():
+			return io.EOF
+		case <-tmr.C:
+		}
+	}
+}
+
+func (vc *vcopier) copyTable(ctx context.Context, tableName string, copyState map[string]*sqltypes.Result) error {
+	defer vc.vr.dbClient.Rollback()
+
+	log.Infof("Copying table %s, lastpk: %v", tableName, copyState[tableName])
+
+	plan, err := buildReplicatorPlan(vc.vr.source.Filter, vc.vr.tableKeys, nil)
+	if err != nil {
+		return err
+	}
+
+	initialPlan, ok := plan.TargetTables[tableName]
+	if !ok {
+		return fmt.Errorf("plan not found for table: %s, curret plans are: %#v", tableName, plan.TargetTables)
+	}
+
+	vsClient, err := tabletconn.GetDialer()(vc.vr.sourceTablet, grpcclient.FailFast(false))
+	if err != nil {
+		return fmt.Errorf("error dialing tablet: %v", err)
+	}
+	defer vsClient.Close(ctx)
+
+	ctx, cancel := context.WithTimeout(ctx, copyTimeout)
+	defer cancel()
+
+	target := &querypb.Target{
+		Keyspace:   vc.vr.sourceTablet.Keyspace,
+		Shard:      vc.vr.sourceTablet.Shard,
+		TabletType: vc.vr.sourceTablet.Type,
+	}
+
+	var lastpkpb *querypb.QueryResult
+	if lastpkqr := copyState[tableName]; lastpkqr != nil {
+		lastpkpb = sqltypes.ResultToProto3(lastpkqr)
+	}
+
+	var pkfields []*querypb.Field
+	var updateCopyState *sqlparser.ParsedQuery
+	err = vsClient.VStreamRows(ctx, target, initialPlan.SendRule.Filter, lastpkpb, func(rows *binlogdatapb.VStreamRowsResponse) error {
+		select {
+		case <-ctx.Done():
+			return io.EOF
+		default:
+		}
+		if vc.tablePlan == nil {
+			if len(rows.Fields) == 0 {
+				return fmt.Errorf("expecting field event first, got: %v", rows)
+			}
+			if err := vc.fastForward(ctx, copyState, rows.Gtid); err != nil {
+				return err
+			}
+			fieldEvent := &binlogdatapb.FieldEvent{
+				TableName: initialPlan.SendRule.Match,
+				Fields:    rows.Fields,
+			}
+			vc.tablePlan, err = plan.buildExecutionPlan(fieldEvent)
+			if err != nil {
+				return err
+			}
+			pkfields = rows.Pkfields
+			buf := sqlparser.NewTrackedBuffer(nil)
+			buf.Myprintf("update _vt.copy_state set lastpk=%a where vrepl_id=%s and table_name=%s", ":lastpk", strconv.Itoa(int(vc.vr.id)), encodeString(tableName))
+			updateCopyState = buf.ParsedQuery()
+		}
+		if len(rows.Rows) == 0 {
+			return nil
+		}
+		query, err := vc.tablePlan.generateBulkInsert(rows)
```

Comment:
It's not immediately clear to me as a reader what factors influence the number of rows that might be included here / the size of the query / the side of the transaction.

Reply:
Added this comment:
```
		// The number of rows we receive depends on the packet size set
		// for the row streamer. Since the packet size is roughly equivalent
		// to data size, this should map to a uniform amount of pages affected
		// per statement. A packet size of 30K will roughly translate to 8
		// mysql pages of 4K each.
```